staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_proto.go

```diff
@@ -38,30 +39,40 @@ func (m *Quantity) Marshal() (data []byte, err error) {
 // MarshalTo is a customized version of the generated Protobuf unmarshaler for a struct
 // with a single string field.
 func (m *Quantity) MarshalTo(data []byte) (int, error) {
-	var i int
+	size := m.Size()
+	return m.LahsramOt(data[:size])
+}
+
+// LahsramOt is a customized version of the generated Protobuf unmarshaler for a struct
+// with a single string field.
+func (m *Quantity) LahsramOt(data []byte) (int, error) {
+	i := len(data)
 	_ = i
 	var l int
 	_ = l
 
-	data[i] = 0xa
-	i++
 	// BEGIN CUSTOM MARSHAL
 	out := m.String()
```

Comment:
Can you check something for me and see whether this allocates a buffer on the heap still?  There should be a benchmark that catches this (either in this package or one of the precanned API object benchmarks for protobuf encode in api) - I just want to know if we're paying the cost here because it might be something to follow up on.

Reply:
I ran this: `go test -benchmem -bench=. -run=NONE ./vendor/k8s.io/apimachinery/pkg/api/resource/`, if that's what you meant:

```
$ benchcmp before after
benchmark                              old ns/op     new ns/op     delta
BenchmarkQuantityString-12             175           169           -3.43%
BenchmarkQuantityStringPrecalc-12      13.9          13.8          -0.72%
BenchmarkQuantityStringBinarySI-12     233           233           +0.00%
BenchmarkQuantityMarshalJSON-12        163           151           -7.36%
BenchmarkQuantityUnmarshalJSON-12      155           154           -0.65%
BenchmarkParseQuantity-12              92.1          88.6          -3.80%
BenchmarkCanonicalize-12               71.1          73.4          +3.23%
BenchmarkQuantityRoundUp-12            15.1          15.4          +1.99%
BenchmarkQuantityCopy-12               106           103           -2.83%
BenchmarkQuantityAdd-12                24.4          24.2          -0.82%
BenchmarkQuantityCmp-12                18.2          18.1          -0.55%
BenchmarkScaledValueSmall-12           29.3          29.5          +0.68%
BenchmarkScaledValueLarge-12           422           416           -1.42%

benchmark                              old allocs     new allocs     delta
BenchmarkQuantityString-12             2              2              +0.00%
BenchmarkQuantityStringPrecalc-12      0              0              +0.00%
BenchmarkQuantityStringBinarySI-12     2              2              +0.00%
BenchmarkQuantityMarshalJSON-12        1              1              +0.00%
BenchmarkQuantityUnmarshalJSON-12      0              0              +0.00%
BenchmarkParseQuantity-12              0              0              +0.00%
BenchmarkCanonicalize-12               0              0              +0.00%
BenchmarkQuantityRoundUp-12            0              0              +0.00%
BenchmarkQuantityCopy-12               1              1              +0.00%
BenchmarkQuantityAdd-12                0              0              +0.00%
BenchmarkQuantityCmp-12                0              0              +0.00%
BenchmarkScaledValueSmall-12           0              0              +0.00%
BenchmarkScaledValueLarge-12           1              1              +0.00%

benchmark                              old bytes     new bytes     delta
BenchmarkQuantityString-12             38            38            +0.00%
BenchmarkQuantityStringPrecalc-12      0             0             +0.00%
BenchmarkQuantityStringBinarySI-12     48            48            +0.00%
BenchmarkQuantityMarshalJSON-12        35            35            +0.00%
BenchmarkQuantityUnmarshalJSON-12      2             2             +0.00%
BenchmarkParseQuantity-12              0             0             +0.00%
BenchmarkCanonicalize-12               3             3             +0.00%
BenchmarkQuantityRoundUp-12            0             0             +0.00%
BenchmarkQuantityCopy-12               64            64            +0.00%
BenchmarkQuantityAdd-12                0             0             +0.00%
BenchmarkQuantityCmp-12                0             0             +0.00%
BenchmarkScaledValueSmall-12           0             0             +0.00%
BenchmarkScaledValueLarge-12           48            48            +0.00%
```

Reply:
Also went ahead and ran `go test -benchmem -bench=. -run=NONE ./pkg/api/testing/`:

```
benchmark                                                                 old ns/op     new ns/op     delta
BenchmarkPodConversion-12                                                 4432          4605          +3.90%
BenchmarkNodeConversion-12                                                3273          3472          +6.08%
BenchmarkReplicationControllerConversion-12                               5600          5331          -4.80%
BenchmarkPodCopy-12                                                       2508          2554          +1.83%
BenchmarkNodeCopy-12                                                      2809          2796          -0.46%
BenchmarkReplicationControllerCopy-12                                     4205          4196          -0.21%
BenchmarkPodDefaulting-12                                                 437           419           -4.12%
BenchmarkEncodeCodecProtobuf-12                                           11513         8241          -28.42%
BenchmarkEncodeCodecFromInternalProtobuf-12                               15171         13003         -14.29%
BenchmarkEncodeProtobufGeneratedMarshal-12                                8330          6188          -25.71%
BenchmarkDecodeCodecToInternalProtobuf-12                                 21978         21988         +0.05%
BenchmarkDecodeIntoProtobuf-12                                            13171         12427         -5.65%
BenchmarkEncodeCodec-12                                                   35009         34713         -0.85%
BenchmarkEncodeCodecFromInternal-12                                       40051         40297         +0.61%
BenchmarkEncodeJSONMarshal-12                                             32782         31155         -4.96%
BenchmarkDecodeCodec-12                                                   87037         85525         -1.74%
BenchmarkDecodeIntoExternalCodec-12                                       79725         79624         -0.13%
BenchmarkDecodeIntoInternalCodec-12                                       83725         81673         -2.45%
BenchmarkDecodeIntoJSON-12                                                93709         94136         +0.46%
BenchmarkDecodeIntoJSONCodecGenConfigFast-12                              37114         37356         +0.65%
BenchmarkDecodeIntoJSONCodecGenConfigCompatibleWithStandardLibrary-12     38612         39392         +2.02%
BenchmarkToUnstructured-12                                                97176         100176        +3.09%
BenchmarkFromUnstructured-12                                              79672         80252         +0.73%
BenchmarkToUnstructuredViaJSON-12                                         118233        114007        -3.57%
BenchmarkFromUnstructuredViaJSON-12                                       211835        205791        -2.85%
BenchmarkToUnstructuredViaJSONIter/compat-12                              126203        129046        +2.25%
BenchmarkToUnstructuredViaJSONIter/default-12                             106798        114387        +7.11%
BenchmarkToUnstructuredViaJSONIter/fastest-12                             106302        100864        -5.12%
BenchmarkFromUnstructuredViaJSONIter/compat-12                            189745        187042        -1.42%
BenchmarkFromUnstructuredViaJSONIter/default-12                           85923         81076         -5.64%
BenchmarkFromUnstructuredViaJSONIter/fastest-12                           69991         71748         +2.51%

benchmark                                                                 old allocs     new allocs     delta
BenchmarkPodConversion-12                                                 11             11             +0.00%
BenchmarkNodeConversion-12                                                8              8              +0.00%
BenchmarkReplicationControllerConversion-12                               14             14             +0.00%
BenchmarkPodCopy-12                                                       10             10             +0.00%
BenchmarkNodeCopy-12                                                      7              7              +0.00%
BenchmarkReplicationControllerCopy-12                                     15             15             +0.00%
BenchmarkPodDefaulting-12                                                 0              0              +0.00%
BenchmarkEncodeCodecProtobuf-12                                           9              9              +0.00%
BenchmarkEncodeCodecFromInternalProtobuf-12                               14             13             -7.14%
BenchmarkEncodeProtobufGeneratedMarshal-12                                7              7              +0.00%
BenchmarkDecodeCodecToInternalProtobuf-12                                 91             91             +0.00%
BenchmarkDecodeIntoProtobuf-12                                            82             82             +0.00%
BenchmarkEncodeCodec-12                                                   40             40             +0.00%
BenchmarkEncodeCodecFromInternal-12                                       45             45             +0.00%
BenchmarkEncodeJSONMarshal-12                                             35             35             +0.00%
BenchmarkDecodeCodec-12                                                   189            189            +0.00%
BenchmarkDecodeIntoExternalCodec-12                                       184            184            +0.00%
BenchmarkDecodeIntoInternalCodec-12                                       189            189            +0.00%
BenchmarkDecodeIntoJSON-12                                                123            123            +0.00%
BenchmarkDecodeIntoJSONCodecGenConfigFast-12                              128            128            +0.00%
BenchmarkDecodeIntoJSONCodecGenConfigCompatibleWithStandardLibrary-12     173            172            -0.58%
BenchmarkToUnstructured-12                                                199            199            +0.00%
BenchmarkFromUnstructured-12                                              178            178            +0.00%
BenchmarkToUnstructuredViaJSON-12                                         305            305            +0.00%
BenchmarkFromUnstructuredViaJSON-12                                       410            410            +0.00%
BenchmarkToUnstructuredViaJSONIter/compat-12                              422            422            +0.00%
BenchmarkToUnstructuredViaJSONIter/default-12                             403            403            +0.00%
BenchmarkToUnstructuredViaJSONIter/fastest-12                             393            394            +0.25%
BenchmarkFromUnstructuredViaJSONIter/compat-12                            550            550            +0.00%
BenchmarkFromUnstructuredViaJSONIter/default-12                           268            268            +0.00%
BenchmarkFromUnstructuredViaJSONIter/fastest-12                           184            184            +0.00%

benchmark                                                                 old bytes     new bytes     delta
BenchmarkPodConversion-12                                                 2685          2685          +0.00%
BenchmarkNodeConversion-12                                                1696          1696          +0.00%
BenchmarkReplicationControllerConversion-12                               3152          3152          +0.00%
BenchmarkPodCopy-12                                                       2064          2064          +0.00%
BenchmarkNodeCopy-12                                                      2256          2256          +0.00%
BenchmarkReplicationControllerCopy-12                                     3000          3000          +0.00%
BenchmarkPodDefaulting-12                                                 4             4             +0.00%
BenchmarkEncodeCodecProtobuf-12                                           2662          2652          -0.38%
BenchmarkEncodeCodecFromInternalProtobuf-12                               4038          4028          -0.25%
BenchmarkEncodeProtobufGeneratedMarshal-12                                1373          1362          -0.80%
BenchmarkDecodeCodecToInternalProtobuf-12                                 7050          7050          +0.00%
BenchmarkDecodeIntoProtobuf-12                                            4536          4536          +0.00%
BenchmarkEncodeCodec-12                                                   4380          4385          +0.11%
BenchmarkEncodeCodecFromInternal-12                                       5800          5799          -0.02%
BenchmarkEncodeJSONMarshal-12                                             4062          4065          +0.07%
BenchmarkDecodeCodec-12                                                   10188         10188         +0.00%
BenchmarkDecodeIntoExternalCodec-12                                       8923          8923          +0.00%
BenchmarkDecodeIntoInternalCodec-12                                       10188         10188         +0.00%
BenchmarkDecodeIntoJSON-12                                                7537          7537          +0.00%
BenchmarkDecodeIntoJSONCodecGenConfigFast-12                              5645          5645          +0.00%
BenchmarkDecodeIntoJSONCodecGenConfigCompatibleWithStandardLibrary-12     6365          6353          -0.19%
BenchmarkToUnstructured-12                                                13793         13793         +0.00%
BenchmarkFromUnstructured-12                                              6648          6648          +0.00%
BenchmarkToUnstructuredViaJSON-12                                         20581         20580         -0.00%
BenchmarkFromUnstructuredViaJSON-12                                       22965         22981         +0.07%
BenchmarkToUnstructuredViaJSONIter/compat-12                              25231         25243         +0.05%
BenchmarkToUnstructuredViaJSONIter/default-12                             19024         19023         -0.01%
BenchmarkToUnstructuredViaJSONIter/fastest-12                             18818         18830         +0.06%
BenchmarkFromUnstructuredViaJSONIter/compat-12                            44136         44128         -0.02%
BenchmarkFromUnstructuredViaJSONIter/default-12                           12235         12227         -0.07%
BenchmarkFromUnstructuredViaJSONIter/fastest-12                           10939         10938         -0.01%
```

Reply:
The most relevant is this:
```
BenchmarkEncodeProtobufGeneratedMarshal-12      8330          6188          -25.71%
```

About 25% speed improvement when serializing a pod.

Reply:
Yeah, allocations being the same was what I was looking for.